turbopack/crates/turbo-tasks-fs/src/watcher.rs

```diff
@@ -87,33 +99,69 @@ impl DiskWatcher {
         Ok(())
     }
 
+    /// Private helper, assumes that the path has already been added to `self.watching`.
     #[cfg(not(any(target_os = "macos", target_os = "windows")))]
     fn start_watching_dir(
         &self,
         watcher: &mut std::sync::MutexGuard<Option<DiskWatcherInternal>>,
         dir_path: &Path,
         root_path: &Path,
     ) -> Result<()> {
-        use anyhow::Context;
+        // HACK: Rewrite NotFound io errors to PathNotFound
+        // This shouldn't ever happen (notify should transform this for us), but even after
+        // https://github.com/notify-rs/notify/pull/611, it seems like there's still some case where
+        // it can occur with inotify on Linux.
+        fn map_notify_err(mut err: notify::Error) -> notify::Error {
+            if let notify::ErrorKind::Io(io_err) = &err.kind {
+                if io_err.kind() == std::io::ErrorKind::NotFound {
+                    err.kind = notify::ErrorKind::PathNotFound;
+                }
+            }
+            err
+        }
 
         if let Some(watcher) = watcher.as_mut() {
             let mut path = dir_path;
-            while let Err(err) = watcher.watch(path, RecursiveMode::NonRecursive) {
-                if path == root_path {
-                    return Err(err).context(format!(
-                        "Unable to watch {} (tried up to {})",
-                        dir_path.display(),
-                        path.display()
-                    ));
+            let err_with_context = |err| {
+                return Err(err).context(format!(
+                    "Unable to watch {} (tried up to {})",
+                    dir_path.display(),
+                    path.display()
+                ));
+            };
+            loop {
+                match watcher
+                    .watch(path, RecursiveMode::NonRecursive)
+                    .map_err(map_notify_err)
+                {
+                    Ok(()) => break,
+                    Err(
+                        err @ notify::Error {
+                            kind: notify::ErrorKind::PathNotFound,
+                            ..
+                        },
+                    ) => {
+                        // The path was probably deleted before we could process the event. That's
+                        // okay, just make sure we're watching the parent directory, so we can know
+                        // if it gets recreated.
+                        // TODO: We should probably also trigger an invalidation if this happens
```

Comment:
That invalidation is trigger by the parent directory.

```suggestion
```

Reply:
Do we know that the parent directory is watched yet?